fms/models/llava_next.py

```diff
@@ -0,0 +1,586 @@
+                image_size=imsize,
+                grid_pinpoints=self.config.image_grid_pinpoints,
+                patch_size=self.config.vision_config.image_size,
+            )
+            for imsize in image_sizes
+        ]
+
+        if pixel_values.dim() == 5:
+            # stacked if input is (batch_size, num_patches, num_channels, height, width)
+            _pixel_values_list = [
+                pix_val[:num_patch]
+                for pix_val, num_patch in zip(pixel_values, image_num_patches)
+            ]
+            pixel_values = torch.cat(_pixel_values_list, dim=0)
+        elif pixel_values.dim() != 4:
+            # otherwise has to be stacked from list of (num_patches, num_channels, height, width)
+            raise ValueError(
+                f"pixel_values of shape {pixel_values.shape}, expect to be of 4 or 5 dimensions"
+            )
+
+        _, _, image_features = self.vision_tower(
+            pixel_values, output_hidden_states=True
+        )
+        if isinstance(self.config.vision_feature_layer, int):
+            selected_image_feature = image_features[self.config.vision_feature_layer]
+        else:
+            hs_pool = [
+                image_features[layer_idx]
+                for layer_idx in self.config.vision_feature_layer
+            ]
+            selected_image_feature = torch.cat(hs_pool, dim=-1)
+
+        if self.config.vision_feature_select_strategy == "default":
+            selected_image_feature = selected_image_feature[:, 1:]
+
+        image_features = self.multi_modal_projector(selected_image_feature)
+        image_features = torch.split(image_features, image_num_patches, dim=0)
+        return image_features
+
+    def pack_image_features(self, image_features, image_sizes, image_newline=None):
+        new_image_features = []
+
+        for image_idx, image_feature in enumerate(image_features):
+            if image_feature.shape[0] > 1:
+                base_image_feature = image_feature[0]
+                image_feature = image_feature[1:]
+                height = width = (
+                    self.config.vision_config.image_size
+                    // self.config.vision_config.patch_size
+                )
+
+                patch_height, patch_width = self.select_best_resolution(
+                    image_sizes[image_idx], self.config.image_grid_pinpoints
+                )
+                num_patch_height = patch_height // self.config.vision_config.image_size
+                num_patch_width = patch_width // self.config.vision_config.image_size
+
+                image_feature = image_feature.view(
+                    num_patch_height, num_patch_width, height, width, -1
+                )
+                image_feature = image_feature.permute(4, 0, 2, 1, 3).contiguous()
+                image_feature = image_feature.flatten(1, 2).flatten(2, 3)
+                image_feature = self.unpad_image(image_feature, image_sizes[image_idx])
```

Comment:
does this also cause a graph break?

Reply:
Interesting, I know [this](https://github.com/foundation-model-stack/foundation-model-stack/blob/85aba0ed2613a3d64df2a04d832788f35d48f11d/fms/models/llava_next.py#L238) does, but the if comparison inside `unpad_image` might also cause one. Will verify with the next pass targeting graph break removal.